CONTRIBUTING.md

```diff
@@ -210,6 +210,124 @@ the "actual" output, then re-run `cargo test` and make sure everything passes.
+lint's query.
+
+##### Templating
+
+We use the `handlebars` crate for writing these templates\. [More information
+more information about the syntax can be found here](https://handlebarsjs.com/guide/#simple-expressions),
+and [here is where `cargo-semver-checks` defines custom helpers
+](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/src/templating.rs).
+
+All fields marked with `@output` in the `query` in `<lint_name>.ron` are available
+to access with `{{name}}` in the `hint_template`, like in the example above.
+
+##### Testing
+
+When the `witness` field is not `None`, `cargo-semver-checks` tests the witness generation
+of the lint similarly to how it tests the `query` itself\. If you run `cargo test` after adding
+a witness for the first time, the test will fail, because it expects a file containing the
+expected witness output.
+
+To solve this, create a file called `test_outputs/witnesses/<lint_name>.output.ron`. Make the
+contents `{}` currently, but we will change this\. Run `cargo test` again\. The test
+`<lint_name>` should fail (because we told it we expect no witnesses to be generated,
+and this is no longer true).
+
+The failed test should output a message like:
+
+```
+---- query::tests_lints::function_missing stdout ----
+--- actual output ---
+{
+    // cut down for readibility
+    "./test_crates/function_missing/": [
+        (
+            filename: "src/lib.rs",
+            begin_line: 1,
+            hint: "use function_missing::will_be_removed_fn;\nwill_be_removed_fn(...);",
+        ),
+        (
+            filename: "src/lib.rs",
+            begin_line: 4,
+            hint: "use function_missing::pub_use_removed_fn;\npub_use_removed_fn(...);",
+        ),
+    ],
+}
+thread 'query::tests_lints::function_missing' panicked at src/query.rs:751:17:
+Witness output for function_missing did not match expected values:
+Differences (-expected|+actual):
+-{}
++{
++    // cut down for readibility
++    "./test_crates/function_missing/": [
++        (
++            filename: "src/lib.rs",
++            begin_line: 1,
++            hint: "use function_missing::will_be_removed_fn;\nwill_be_removed_fn(...);",
++        ),
++        (
++            filename: "src/lib.rs",
++            begin_line: 4,
++            hint: "use function_missing::pub_use_removed_fn;\npub_use_removed_fn(...);",
++        ),
++    ],
++}
+
+Update the `test_outputs/witnesses/function_missing.output.ron` file if
+                    the new test results are correct.
+```
+
+Check the actual result under `--- actual ---` and make sure that it outputted
+the correct witness hints\. Note that it may contain output for other test crates - this
+not necessarily an error: see the troubleshooting section for more info\. Once you've
+verified that the `--- actual ---` results are as expected, copy the actual results
+into the `test_outputs/witnesses/<lint_name>.output.ron` file and save\. Running
```

Comment:
Is there a viable way to `insta` this? 🤔

Reply:
Totally! I was matching the behavior of the basic lint tests, but we could definitely use `insta` here, and it would probably be better for contributor experience (but less consistent)

Reply:
I imagine we'll eventually move the existing non-insta steps to insta, so probably best to start standardizing toward the new behavior we'll want eventually anyway. If someone is annoyed at the inconsistency, we can tell them we'd love a PR to move everything to insta 😅

Reply:
OK, I'll open a PR to use insta for the tests instead, then update the docs here.

Reply:
#935